### PR TITLE
#82854 Implement GET Subscriber API

### PR DESCRIPTION
**Pull request by tabman83**

### CODE REVIEW

File: src/CaptainHook.Api/Controllers/EventsController.cs [hunk 1]
```diff
@@ -183,5 +188,47 @@ public async Task<IActionResult> DeleteSubscriber([FromRoute] string eventName,
...
+        {
+            try
+            {
+                var directorServiceUri = new Uri(ServiceNaming.DirectorServiceFullName);
+                var directorServiceClient = ServiceProxy.Create<IDirectorServiceRemoting>(directorServiceUri);
+                var subscribers = await directorServiceClient.GetAllSubscribersAsync();
+
+                if (subscribers.Count == 0)
+                {
+                    return StatusCode(StatusCodes.Status503ServiceUnavailable);
```
Comment (chomick) [inline src/CaptainHook.Api/Controllers/EventsController.cs:215]: Is that what we return today? I would simply return 404 - what if you have no subscribers? why 503?

File: src/CaptainHook.Api/Controllers/EventsController.cs [hunk 2]
```diff
@@ -183,5 +188,47 @@ public async Task<IActionResult> DeleteSubscriber([FromRoute] string eventName,
...
+                }
+
+                if (!subscribers.TryGetValue($"{eventName};{subscriberName}", out SubscriberConfiguration subscriberConfiguration))
```
Comment (chomick) [inline src/CaptainHook.Api/Controllers/EventsController.cs:218]: Is the way the key is generated something that can be extracted/shared? 
Also, you will have callback and DLQ cases as well - the key generation algorithm should be shared here I think

File: src/CaptainHook.Api/Controllers/EventsController.cs [hunk 3]
```diff
@@ -183,5 +188,47 @@ public async Task<IActionResult> DeleteSubscriber([FromRoute] string eventName,
...
+                {
+                    return NotFound();
+                }
+
+                AuthenticationConfigSanitizer.Sanitize(subscriberConfiguration);
+
+                return Ok(subscriberConfiguration);
```
Comment (chomick) [inline src/CaptainHook.Api/Controllers/EventsController.cs:225]: Here you may have more than 1 entry: dlq, callback as well are considered separate subscribers.
Unless you want to have independent endpoint for getting those?

**Review by chomick — COMMENTED** [review]
DLQ/Callback support missing.
Will you add unit tests for the controller & integration tests for security?
Also,. integration tests could be updated to use this method rather than getAll for some verifications.

File: src/CaptainHook.Api/Controllers/EventsController.cs [hunk 4]
```diff
@@ -183,5 +188,47 @@ public async Task<IActionResult> DeleteSubscriber([FromRoute] string eventName,
...
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status503ServiceUnavailable)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> GetSubscriber([FromRoute] string eventName, [FromRoute] string subscriberName)
+        {
+            try
+            {
+                var directorServiceUri = new Uri(ServiceNaming.DirectorServiceFullName);
```
Comment (dawidkeshop) [inline src/CaptainHook.Api/Controllers/EventsController.cs:209]: I don't think it's a good idea to make such inconsistency by not following our architecture. We should not call DirectorService directly from controllers, this should be placed in an Application Handler.

File: src/CaptainHook.Fabric/ApplicationParameters/Local.1Node.xml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Application xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="fabric:/CaptainHook" xmlns="http://schemas.microsoft.com/2011/01/fabric">
   <Parameters>
-    <Parameter Name="KeyVault_Base_Uri" Value="https://esw-tooling-ci-we.vault.azure.net/" />
+    <Parameter Name="KeyVault_Base_Uri" Value="https://esw-tooling-ci-we-ap.vault.azure.net/" />
```
Comment (dawidkeshop) [inline src/CaptainHook.Fabric/ApplicationParameters/Local.1Node.xml:4]: I don't think this is a good idea.

**Review by dawidkeshop — CHANGES_REQUESTED** [review]

File: src/CaptainHook.Fabric/ApplicationParameters/Local.1Node.xml (hunk 5 above)
Comment (tabman83) [reply src/CaptainHook.Fabric/ApplicationParameters/Local.1Node.xml:4]: That slipped through, thanks

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 4 above)
Comment (tabman83) [reply src/CaptainHook.Api/Controllers/EventsController.cs:209]: I can implement handler here but it's gonna grow more than the original estimation of course. I followed the original story description of not touching too many layers for this. @chomik?

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 1 above)
Comment (tabman83) [reply src/CaptainHook.Api/Controllers/EventsController.cs:215]: 503 represents director service not ready yet (busy loading configuration).
It was initially returned when subscribers was null, but a different situation prompted to have the subscribers collection to be initialized on the director side. So this became broken and checking the count was a quick solution to get it back working.

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 4 above)
Comment (chomick) [reply src/CaptainHook.Api/Controllers/EventsController.cs:209]: @dawidkeshop 
We are replicating the existing GET functionality , just adding filtering to improve usage.
These endpoints are going to be reimplemented as they do not return correct contracts.
I want to minimise effort on those so I am happy to follow the same approach we did for first GET.

When/if we implement the GET properly, we will follow the correct pattern/arch.

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 3 above)
Comment (tabman83) [reply src/CaptainHook.Api/Controllers/EventsController.cs:225]: Callback is part of the subscriber configuration.
Dlq is a subscriber per se, so you just use 'dlq' as a subscriber name to get it

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 2 above)
Comment (tabman83) [reply src/CaptainHook.Api/Controllers/EventsController.cs:218]: we get DLQ through the subscriber name, see below

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 4 above)
Comment (dawidkeshop) [reply src/CaptainHook.Api/Controllers/EventsController.cs:209]: So it was my misunderstanding of the scope. Apologies.

**Review by dawidkeshop — APPROVED** [review]

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 2 above)
Comment (chomick) [reply src/CaptainHook.Api/Controllers/EventsController.cs:218]: As discussed, let's keep it simple - the subscriber name (different from subscriber name used in the PUT operations) will have to be explicit. We also agreed to add a very simple endpoint to list names of existing subscribers (KV config models) so they can be queried by the GET for details.

**Review by chomick — APPROVED** [review]

**Pull request merged**